### PR TITLE
Support `qml.counts` for circuits

### PR DESCRIPTION
**Pull request by king-p3nguin**
*Issue #, if available:*
#195 

*Description of changes:*
- `qml.counts()` is now supported.

*Testing done:*

Example:
```python
import pennylane as qml
dev = qml.device("braket.local.qubit", wires=2, shots=100)

@qml.qnode(dev)
def test_circ():
    qml.PauliX(0)
    return qml.counts()

test_circ()
```

output:
```text
tensor(Counter({'10': 100}), dtype=object, requires_grad=True)
```

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yitchen-tim** [comment]
Hi @king-p3nguin, thanks for starting your contribution! Could you also add unit tests to this PR? When this PR is ready, you can flip it from draft PR to ready for review.

**Comment by yitchen-tim** [comment]
Hi @king-p3nguin, the basic use cases from the [issue](https://github.com/amazon-braket/amazon-braket-pennylane-plugin-python/issues/195) throws error
```
import pennylane as qml
dev = qml.device("braket.local.qubit", wires=2, shots=100)

@qml.qnode(dev)
def test_circ():
    qml.PauliZ(0)
    return qml.counts()

test_circ()
```
with error message "DeviceError: Unsupported observable: <class 'NoneType'>"

And another use case 
```
@qml.qnode(dev)
def circuit():
    qml.X(wires=0)
    return qml.counts(op=qml.PauliZ(wires=[0]))
```
throws a different error `NotImplementedError: Unsupported return type: ObservableReturnTypes.Counts`

**Comment by yitchen-tim** [comment]
Also, could you add more tests to increase the coverage? thanks

**Comment by king-p3nguin** [comment]
@yitchen-tim Thank you for checking. The two basic use cases did not show errors in my environment. Can you ensure the code is executed within my fork's `counts` branch? (The added integrated tests have a similar circuit, and those tests are working fine on CI)

**Comment by yitchen-tim** [comment]
@king-p3nguin  Great, thanks, I did have some dependencies so it was not working correctly. The basic examples works for me now. After full test coverage, we should be good to go! 

Also, per UnitaryHack guideline, it's good as long as PR is opened before EOD today and merged before 6/19. So we still have a week.

File: src/braket/pennylane_plugin/translation.py [hunk 1]
```diff
@@ -558,22 +561,29 @@ def translate_result_type(
...
     if return_type is ObservableReturnTypes.Expectation:
         return Expectation(braket_observable, targets)
     elif return_type is ObservableReturnTypes.Variance:
         return Variance(braket_observable, targets)
-    elif return_type is ObservableReturnTypes.Sample:
+    elif return_type in (ObservableReturnTypes.Sample, ObservableReturnTypes.Counts):
         return Sample(braket_observable, targets)
     else:
-        raise NotImplementedError(f"Unsupported return type: {return_type}")
+        raise NotImplementedError(f"Unsupported return type: {return_type}")  # pragma: no cover
```
Comment (yitchen-tim) [inline src/braket/pennylane_plugin/translation.py:586]: Can you add test case for unsupported return type rather than ` # pragma: no cover`?

**Comment by king-p3nguin** [comment]
@yitchen-tim 
Sorry for the delay. I added unit tests.

File: src/braket/pennylane_plugin/translation.py (hunk 1 above)
Comment (king-p3nguin) [reply src/braket/pennylane_plugin/translation.py]: It seems that when an observable is defined, there are only four possible types (`Expectation`, `Variance`, `Sample`, and `Counts`), so I will just use the else statement here.

Comment (yitchen-tim) [reply src/braket/pennylane_plugin/translation.py]: The else statement is prone to error. Pennylane supports [few other return types](https://github.com/PennyLaneAI/pennylane/blob/d616c3e56e6bbc95cfbeec6c1f949b6e822caa8d/pennylane/measurements/measurements.py#L39), and the else statement doesn't really handle all these other result types. 

To add coverage and test the unsupported return type for these lines
```
    else:
        raise NotImplementedError(f"Unsupported return type: {return_type}")
```
, you could mock the `measurement` as something like
```
measurement = Mock(
    return_type = Enum('ObservableReturnTypes', {'Foo': "foo"}).Foo,
    observable = qml.PauliZ()
)
```
If any function that is imcompatible with the mocked `measurement`, you can adjust the mock or path the imcompatible function.

**Comment by yitchen-tim** [comment]
@king-p3nguin thanks, coverage is 100%, woot! Just one more change around testing unsupported result types.

File: src/braket/pennylane_plugin/translation.py (hunk 1 above)
Comment (king-p3nguin) [reply src/braket/pennylane_plugin/translation.py]: Thank you for the suggestion! I made the change as requested.

**Review by yitchen-tim — APPROVED** [review]
Great work! Thanks again for the contribution @king-p3nguin

**Review by rmshaffer — APPROVED** [review]

**Pull request merged**